File: ICPC/WarmUp 10/ncpc2015solutions/floppy_lukas.cs
//Solution by lukasP (Lukáš Poláček)
using System;
using System.IO;
using System.Collections;
using System.Linq;

class Problem
{
    public static void Main ()
    {
        var bs = new BufferedStream(Console.OpenStandardInput());
        StreamReader reader = new StreamReader(bs);

        foreach (int cas in Enumerable.Range(0, int.Parse(reader.ReadLine())))
        {
            int[] w = reader.ReadLine().Split().Select(int.Parse).ToArray();
            int t = w[0], n = w[1];
            BitArray can = new BitArray(t + 1, true);

            foreach (int i in Enumerable.Range(0, n))
            {
                w = reader.ReadLine().Split().Select(int.Parse).ToArray();
                int start = w[0], end = w[1];
                int dif = end - start;

                BitArray ncan = new BitArray(t + 1, false);
                for (int j = 0; j < t + 1; j++) if (can[j])
                {
                    if (j >= dif)
                        ncan[j - dif] = true;
                    if (j + dif <= t)
                        ncan[j + dif] = true;
                }
                can = ncan;
            }

            if (!(from bool m in can where m select m).Any())
            {
                Console.WriteLine("impossible");
                return;
            }
        }
        Console.WriteLine("possible");
    }
}

File: ICPC/WarmUp 10/ncpc2015solutions/floppy_lukas.cpp
//Solution by lukasP (Lukáš Poláček)
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdio>
using namespace std;

#define rep(i,a,b) for(__typeof(b) i=a; i<(b); ++i)

int main()
{
    int f; scanf("%d", &f);
    rep(cas,0,f)
    {
        int t, n;
        scanf("%d %d", &t, &n);
        vector<bool> can(t + 1, true);
        vector<bool> ncan = can;

        rep(i,0,n)
        {
            int start, end;
            scanf("%d %d", &start, &end);
            const int dif = end - start;

            fill(ncan.begin(), ncan.end(), 0);
            rep(j,0,t+1) if (can[j])
            {
                if (j >= dif)
                    ncan[j - dif] = true;
                if (j + dif <= t)
                    ncan[j + dif] = true;
            }
            swap(ncan, can);
        }
        auto is_false = [](bool const x) { return !x; };
        if (all_of(can.begin(), can.end(), is_false))
        {
            cout << "impossible" << endl;
            return 0;
        }
    }
    cout << "possible" << endl;
}
